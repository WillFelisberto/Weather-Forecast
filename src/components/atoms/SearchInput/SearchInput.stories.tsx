/* eslint-disable no-console */
import type { Meta, StoryObj } from '@storybook/react';

import { SearchInput } from './SearchInput';

const meta: Meta<typeof SearchInput> = {
  title: 'Atoms/SearchInput',
  component: SearchInput,
  tags: ['autodocs'],
  args: {
    placeholder: 'Search here...'
  }
};

export default meta;

type Story = StoryObj<typeof SearchInput>;

export const Default: Story = {
  args: {
    onSearchCallback: async (query) => {
      console.log(`Searching for: ${query}`);
    }
  }
};

export const WithCustomPlaceholder: Story = {
  args: {
    placeholder: 'Type something...',
    onSearchCallback: async (query) => {
      console.log(`Searching for: ${query}`);
    }
  }
};

export const Focused: Story = {
  args: {
    placeholder: 'Start typing...',
    onSearchCallback: async (query) => {
      console.log(`Searching for: ${query}`);
    }
  },
  play: ({ canvasElement }) => {
    const input = canvasElement.querySelector('input');
    input?.focus();
  }
};
