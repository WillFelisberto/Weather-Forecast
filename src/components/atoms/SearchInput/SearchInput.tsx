import clsx from 'clsx';
import { Loader2, Search } from 'lucide-react';
import { useState, KeyboardEvent } from 'react';

import { Button } from '../Button';

interface SearchInputProps {
  placeholder?: string;
  className?: string;
  onSearchCallback(query: string): Promise<void>;
}

export const SearchInput: React.FC<SearchInputProps> = ({
  placeholder = 'Search here...',
  className,
  onSearchCallback
}) => {
  const [searchValue, setSearchValue] = useState('');
  const [isSearching, setIsSearching] = useState(false);

  const handleSearch = async () => {
    if (searchValue.trim()) {
      setIsSearching(true);
      try {
        await onSearchCallback(searchValue);
      } finally {
        setIsSearching(false);
      }
    }
  };

  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      handleSearch();
    }
  };

  return (
    <div className="flex items-center gap-2">
      <input
        type="text"
        value={searchValue}
        onChange={(e) => setSearchValue(e.target.value)}
        onKeyDown={handleKeyDown}
        placeholder={placeholder}
        className={clsx(
          'flex-1 px-4 py-2 rounded-[12px] text-white bg-[--primary-color] placeholder-gray-400 focus:outline-none focus:ring-2 focus:ring-blue-500 transition-all',
          className
        )}
        disabled={isSearching}
      />
      <Button
        onClick={handleSearch}
        disabled={isSearching}
        className={clsx(
          'px-4 py-2  rounded-md text-white transition-colors bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500',
          { 'opacity-50 cursor-not-allowed': isSearching }
        )}
      >
        {isSearching ? (
          <Loader2 role="search-spin" className="animate-spin" size={20} />
        ) : (
          <Search role="search-icon" size={20} />
        )}
      </Button>
    </div>
  );
};
