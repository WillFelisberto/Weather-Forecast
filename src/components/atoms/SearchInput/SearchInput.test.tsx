import { fireEvent, render, screen } from '@/tests/test-utils';
import '@testing-library/jest-dom';

import { SearchInput } from './SearchInput';

describe('SearchInput Component', () => {
  it('renders the input with placeholder', () => {
    const { container } = render(
      <SearchInput placeholder="Test search..." onSearchCallback={jest.fn()} />
    );
    expect(screen.getByPlaceholderText('Test search...')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('triggers the search callback when Enter is pressed', async () => {
    const mockCallback = jest.fn().mockResolvedValue(undefined);
    render(<SearchInput onSearchCallback={mockCallback} />);

    const input = screen.getByPlaceholderText('Search here...');
    fireEvent.change(input, { target: { value: 'San Francisco' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });

    expect(mockCallback).toHaveBeenCalledWith('San Francisco');
  });

  it('displays the spinner and disables the button during search', async () => {
    const mockCallback = jest
      .fn()
      .mockImplementation(() => new Promise((resolve) => setTimeout(resolve, 1000)));
    render(<SearchInput onSearchCallback={mockCallback} />);

    const input = screen.getByPlaceholderText('Search here...');
    const button = screen.getByRole('button');

    fireEvent.change(input, { target: { value: 'Test' } });
    fireEvent.click(button);

    expect(button).toBeDisabled();
    expect(screen.getByRole('search-spin', { hidden: true })).toBeInTheDocument(); // Spinner
  });
});
