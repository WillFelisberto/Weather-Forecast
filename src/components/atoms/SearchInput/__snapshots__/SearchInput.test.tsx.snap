// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Component renders the input with placeholder 1`] = `
<div>
  <div
    class="flex items-center gap-2"
  >
    <input
      class="flex-1 px-4 py-2 rounded-[12px] text-white bg-[--primary-color] placeholder-gray-400 focus:outline-none focus:ring-2 focus:ring-blue-500 transition-all"
      placeholder="Test search..."
      type="text"
      value=""
    />
    <button
      class="px-4 py-2 rounded-xl text-white font-medium transition-all bg-gradient-to-r from-[#3A5E72] to-[#5C93B1] hover:opacity-80 px-4 py-2  rounded-md text-white transition-colors bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500"
    >
      <svg
        class="lucide lucide-search"
        fill="none"
        height="20"
        role="search-icon"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
    </button>
  </div>
</div>
`;
